x-pack/plugins/actions/server/lib/action_executor.ts

```diff
@@ -328,55 +327,46 @@ export class ActionExecutor {
           eventLogger.logEvent(event);
         }
 
-        // start openai extension
-        // add event.kibana.action.execution.openai to event log when OpenAI Connector is executed
-        if (result.status === 'ok' && actionTypeId === '.gen-ai') {
-          const data = result.data as unknown as {
-            usage: { prompt_tokens?: number; completion_tokens?: number; total_tokens?: number };
-          };
-          event.kibana = event.kibana || {};
-          event.kibana.action = event.kibana.action || {};
-          event.kibana = {
-            ...event.kibana,
-            action: {
-              ...event.kibana.action,
-              execution: {
-                ...event.kibana.action.execution,
-                gen_ai: {
-                  usage: {
-                    total_tokens: data.usage?.total_tokens,
-                    prompt_tokens: data.usage?.prompt_tokens,
-                    completion_tokens: data.usage?.completion_tokens,
+        // start genai extension
+        if (result.status === 'ok') {
+          getGenAiTokenTracking({
+            actionTypeId,
+            logger,
+            result,
+            validatedParams,
+          })
+            .then((tokenTracking) => {
+              if (tokenTracking != null) {
+                event.kibana = event.kibana || {};
+                event.kibana.action = event.kibana.action || {};
+                event.kibana = {
+                  ...event.kibana,
+                  action: {
+                    ...event.kibana.action,
+                    execution: {
+                      ...event.kibana.action.execution,
+                      gen_ai: {
+                        usage: {
+                          total_tokens: tokenTracking.total_tokens,
+                          prompt_tokens: tokenTracking.prompt_tokens,
+                          completion_tokens: tokenTracking.completion_tokens,
+                        },
+                      },
+                    },
```

Comment:
NIT: we could simplify this a little bit using `import { set } from '@kbn/safer-lodash-set';`
```suggestion
               set(event, 'kibana.action.execution.get_ai.usage', 
                   {
                       total_tokens: tokenTracking.total_tokens,
                       prompt_tokens: tokenTracking.prompt_tokens,
                       completion_tokens: tokenTracking.completion_tokens,
                   }
               );
```
It will also create missing objects.